Reworked registration form and register_user logic

kuba-lilz opened the pull request:
This is a work in progress PR meant to discuss how to go about implementing https://github.com/jwag956/flask-security/issues/251

Code in here only changes internals, and shouldn't affect any clients.
It adds a way to extract both user model data and all form data from registration form.

The next step would be to expose form data to clients and I would like to ask for ideas on how to do that. Adding that data as say `form_data` argument to `user_registered` signal seems most natural. But it would break clients that use this event - as it would require them to add new argument to their `user_registered` callbacks. Is that fine?
```python
form_data = registration_form.to_dict(only_user=False)
user_registered.send(app._get_current_object(), user=user, confirm_token=token, registration_form_data=form_data)
```

Any preferences/ideas on how to expose form_data without affecting existing clients? Adding a new signal seems a bit silly.

jwag956 commented on flask_security/views.py:
```diff
@@ -216,7 +216,7 @@ def register():
     if form.validate_on_submit():
         did_login = False
-        user = register_user(**form.to_dict())
+        user = register_user(registration_form=form)
```
no reason to make this a keyword arg - just pass 'form'

jwag956 commented on flask_security/forms.py:
```diff
@@ -212,13 +212,34 @@ def validate_next(self, field):
+            is_form_field = isinstance(member, Field)
+
+            # Not a form field, return False
+            if is_form_field is False:
```
common usage is:

```
if not is_form_field:
  return False
```
lets not introduce tons of superfluous white space!

jwag956 commented:
I think adding an extra argument is fine - users are supposed to define their signal handlers with **extra - flask specifically mentions this case of adding more info:

http://flask.palletsprojects.com/en/1.1.x/signals/

kuba-lilz commented:
Refactored comments as requested.
I also modified `user_registered` to include `form_data` dictionary - with all fields, though if you would like to filter out user fields from there, that can be done.

There was one test (`tests/test_registerable.py::test_registerable_flag`) I had to modify, but I only added an assert that form_data isn't empty.
A test with some form including specific user-defined input asserting against it would be best, but I didn't find any example test that would expose registration form and go all the way to registering user. Can you point me out to something I can work off?

I also haven't modified docs yet.

kuba-lilz commented:
Added tests/test_registerable.py::test_form_data_is_passed_to_user_registered_signal

kuba-lilz commented:
I think docs should be updated too - should I add docs update to this PR or split into a separate one?

jwag956 commented:
Yes - please do - for something 'small' like this - it is best to have a single/PR. Once you are 'done' - remove the WIP and I will review again.
Be sure to run the build_sphinx locally to catch any errors.

kuba-lilz commented:
Added changes to api documentation - I believe that's the only part that references any information about user_registered callback. Personally I think it would be nice to explain a bit more what each argument sent to callback means, but I kept explanation terse, in spirit with current docs.

A minor detail - keyword I added to user registration signal is `form_data`. When registering user with json request from user's point of view there is no form involved, so maybe `registration_data` would be more appropriate. On the other hand even though user doesn't go through registration form, flask security passes json through registration form, so in that sense `form_data` is appropriate. 
If you think that maybe `registration_data` or a different name would work better, please let me know.

jwag956 commented on docs/api.rst:
```diff
@@ -133,7 +133,7 @@ sends the following signals.
 
    Sent when a user registers on the site. In addition to the app (which is the
-   sender), it is passed `user` and `confirm_token` arguments.
+   sender), it is passed `user`, `confirm_token` and `form_data` arguments.
```
I think it would be useful as you mentioned to add a sentence about form_data - that is is all the Fields from the RegistrationForm. I think the name is fine since even with JSON - a user has to define a MyRegistrationForm....

kuba-lilz commented:
Sure, I added a bit more info, how about now?
